tests/test_ops/test_rans.py

```diff
@@ -0,0 +1,123 @@
+# Copyright (c) OpenMMLab. All rights reserved.
+import pickle
+import time
+
+import pytest
+import torch
+
+from mmcv.ops import (pmf_to_quantized_cdf, rans_decode_with_indexes,
+                      rans_encode_with_indexes)
+
+EXTEND_TIMES = 100
+DEVICES = ['cpu', 'cuda'] if torch.cuda.is_available() else ['cpu']
+
+
+@pytest.fixture(scope='module')
+def rans_input_data():
+    inputs = torch.load('tests/data/for_rans/rans_encoder_inputs.pt')
+    symbols = inputs['symbols'].repeat(EXTEND_TIMES)
+    indexes = inputs['indexes'].repeat(EXTEND_TIMES)
+    cdfs = inputs['quantized_cdfs']
+    cdfs_sizes = inputs['cdfs_sizes']
+    offsets = inputs['offsets']
+    return symbols, indexes, cdfs, cdfs_sizes, offsets
+
+
+@pytest.fixture(scope='module')
+def rans_output_data():
+    with open('tests/data/for_rans/rans_encoder_output.pkl', 'rb') as f:
+        encoded = pickle.load(f)
+    return encoded
+
+
+@pytest.fixture(scope='module')
+def pmf_input_data():
+    inputs = torch.load('tests/data/for_rans/pmf_to_quantized_cdf_inputs.pt')
+    pmfs = inputs['pmfs']
+    pmf_lengths = inputs['pmf_lengths']
+    tail_masses = inputs['tail_masses']
+    return pmfs, pmf_lengths, tail_masses
+
+
+@pytest.fixture(scope='module')
+def pmf_output_data():
+    output = torch.load('tests/data/for_rans/pmf_to_quantized_cdf_output.pt')
+    quantized_cdfs = output['quantized_cdfs']
+    return quantized_cdfs
+
+
+def _move_data_to_target_device(data, device):
+    if isinstance(data, torch.Tensor):
+        return data.to(device)
+    return [d.to(device) for d in data]
+
+
+def _calculate_mibs(fn, device, num_threads, num_symbols, args):
+    if torch.cuda.is_available():
+        torch.cuda.synchronize()
+    start_time = time.time()
+    for _ in range(100):
+        fn(*args)
+    if torch.cuda.is_available():
+        torch.cuda.synchronize()
+    end_time = time.time()
+    latency = (end_time - start_time) / 100
+    mibs = num_symbols / latency / 1024 / 1024
+    print(f'\ndevice {device}, num_threads {num_threads}: \
+            {fn.__name__}: {latency:.4f}s, {mibs:.2f}MiB/s')
+
+
+@pytest.mark.parametrize('device', DEVICES)
+@pytest.mark.parametrize('num_threads', [1, 2, 4])
```

Comment:
Test failed with num_threads > 8 on my device. Is there any limit on the device? can we detect it and set the num_threads according to the environment?

Reply:
We add max cpu threds limitation.
```python
NUM_THREDS = [i for i in [1, 2, 4] if i <= cpu_count()]
```